File: springboot-backend/src/main/kotlin/com/yyjy/memo/controller/PostController.kt
package com.yyjy.memo.controller

import cn.dev33.satoken.annotation.SaCheckLogin
import com.yyjy.memo.common.PageRes
import com.yyjy.memo.common.R
import com.yyjy.memo.models.entity.Posts
import com.yyjy.memo.models.entity.by
import com.yyjy.memo.models.entity.dto.NoteCreateInput
import com.yyjy.memo.models.entity.dto.PhotoCreateInput
import com.yyjy.memo.service.PostService
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.tags.Tag
import org.babyfish.jimmer.Page
import org.babyfish.jimmer.client.FetchBy
import org.babyfish.jimmer.client.meta.Api
import org.babyfish.jimmer.sql.kt.fetcher.newFetcher
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*

@Api
@Tag(name = "帖子模块")
@RequestMapping("posts")
@RestController
class PostController(
	private val postService: PostService
) {

	@Api
	@Operation(summary = "分页获取帖子列表")
	@GetMapping("/")
	fun getNoteList(limit: Int, size: Int, categoryId: Long?): PageRes<@FetchBy("POST_ITEM") Posts> {
		val page = postService.notePage(limit, size, categoryId, POST_ITEM)
		return PageRes(page.totalRowCount, limit, size, page.rows)
	}

	@Api
	@Operation(summary = "获取详情")
	@GetMapping("/{id}")
	fun getDetail(@PathVariable("id") id: Long) {
//		postService.getDetail(id)
		TODO()
	}

	@Api
	@SaCheckLogin
	@Operation(summary = "发布便签")
	@PostMapping("/")
	fun postNote(@Validated @RequestBody note: NoteCreateInput): R<String?> {
		postService.postNote(note)
		return R.ok()
	}

	@Api
	@SaCheckLogin
	@Operation(summary = "发布照片")
	@PostMapping("/upload")
	fun postPhoto(@Validated @RequestBody photo: PhotoCreateInput): R<String?> {
		postService.postPhoto(photo)
		return R.ok()
	}

	@Api
	@SaCheckLogin
	@Operation(summary = "删除内容")
	@DeleteMapping("/{id}")
	fun delete(@PathVariable("id") postId: Long): R<String?> {
		postService.delete(postId)
		return R.ok()
	}

	@Api
	@Operation(summary = "根据用户名搜索便签")
	@GetMapping("/search")
	fun search(username: String, limit: Int, size: Int): PageRes<@FetchBy("POST_ITEM") Posts> {
		val page: Page<Posts> = postService.pageByUsername(username, limit, size, POST_ITEM)
		return PageRes(page.totalRowCount, limit, size, page.rows)
	}

	companion object {
		val POST_ITEM  = newFetcher(Posts::class).by {
			category {
				name()
			}
			content()
			imageUrl()
			cardColor()
			user {
				username()
			}
			likeCount()
			commentCount()
			createdAt()
		}
	}
}

File: react-admin/src/__generated/services/PostController.ts
import type {Executor} from '../';
import type {PostsDto} from '../model/dto/';
import type {
    NoteCreateInput, 
    PageRes, 
    PhotoCreateInput, 
    R
} from '../model/static/';

export class PostController {
    
    constructor(private executor: Executor) {}
    
    readonly delete: (options: PostControllerOptions['delete']) => Promise<
        R<string | undefined>
    > = async(options) => {
        let _uri = '/posts/';
        _uri += encodeURIComponent(options.postId);
        return (await this.executor({uri: _uri, method: 'DELETE'})) as Promise<R<string | undefined>>;
    }
    
    readonly getDetail: (options: PostControllerOptions['getDetail']) => Promise<
        void
    > = async(options) => {
        let _uri = '/posts/';
        _uri += encodeURIComponent(options.id);
        return (await this.executor({uri: _uri, method: 'GET'})) as Promise<void>;
    }
    
    readonly getNoteList: (options: PostControllerOptions['getNoteList']) => Promise<
        PageRes<PostsDto['PostController/POST_ITEM']>
    > = async(options) => {
        let _uri = '/posts/';
        let _separator = _uri.indexOf('?') === -1 ? '?' : '&';
        let _value: any = undefined;
        _value = options.limit;
        _uri += _separator
        _uri += 'limit='
        _uri += encodeURIComponent(_value);
        _separator = '&';
        _value = options.size;
        _uri += _separator
        _uri += 'size='
        _uri += encodeURIComponent(_value);
        _separator = '&';
        _value = options.categoryId;
        _uri += _separator
        _uri += 'categoryId='
        _uri += encodeURIComponent(_value);
        _separator = '&';
        return (await this.executor({uri: _uri, method: 'GET'})) as Promise<PageRes<PostsDto['PostController/POST_ITEM']>>;
    }
    
    readonly postNote: (options: PostControllerOptions['postNote']) => Promise<
        R<string | undefined>
    > = async(options) => {
        let _uri = '/posts/';
        return (await this.executor({uri: _uri, method: 'POST', body: options.body})) as Promise<R<string | undefined>>;
    }
    
    readonly postPhoto: (options: PostControllerOptions['postPhoto']) => Promise<
        R<string | undefined>
    > = async(options) => {
        let _uri = '/posts/upload';
        return (await this.executor({uri: _uri, method: 'POST', body: options.body})) as Promise<R<string | undefined>>;
    }
    
    readonly search: (options: PostControllerOptions['search']) => Promise<
        PageRes<PostsDto['PostController/POST_ITEM']>
    > = async(options) => {
        let _uri = '/posts/search';
        let _separator = _uri.indexOf('?') === -1 ? '?' : '&';
        let _value: any = undefined;
        _value = options.username;
        _uri += _separator
        _uri += 'username='
        _uri += encodeURIComponent(_value);
        _separator = '&';
        _value = options.limit;
        _uri += _separator
        _uri += 'limit='
        _uri += encodeURIComponent(_value);
        _separator = '&';
        _value = options.size;
        _uri += _separator
        _uri += 'size='
        _uri += encodeURIComponent(_value);
        _separator = '&';
        return (await this.executor({uri: _uri, method: 'GET'})) as Promise<PageRes<PostsDto['PostController/POST_ITEM']>>;
    }
}

export type PostControllerOptions = {
    'getNoteList': {
        readonly limit: number, 
        readonly size: number, 
        readonly categoryId: number
    }, 
    'getDetail': {
        readonly id: number
    }, 
    'postNote': {
        readonly body: NoteCreateInput
    }, 
    'postPhoto': {
        readonly body: PhotoCreateInput
    }, 
    'delete': {
        readonly postId: number
    }, 
    'search': {
        readonly username: string, 
        readonly limit: number, 
        readonly size: number
    }
}
